### PR TITLE
Cmake support for models

### PR DESCRIPTION
**Pull request by patstew**
This is the CMake support from issue #1363 with mine and Maarten's changes rebased to master

### CODE REVIEW

**Comment by madebr** [comment]
@patstew, I guess these commits are a copy of the cmake branch of my fork? Then LGTM.

**Comment by madebr** [comment]
@patstew I haven't tested systemc yet, but I have rebased this on current master in my [patstew branch](https://github.com/madebr/verilator/tree/patstew_cmake)

**Comment by patstew** [comment]
I've just done the same :)

**Comment by madebr** [comment]
Main diffs are:
- subdirs in `Makefile.in`
- added `cmake_hello_sc`
- Added header(s) + comments to `examples/cmake_hello_c/CMakeLists.txt` and `examples/cmake_hello_c/Makefile`
- Simpler `examples/cmake_tracing_c/.gitignore`
- Some examples have more descriptive `DESCRIPTION`.

But all in general small.
I have tested `make examples` with systemc and confirm everything is working.

**Comment by patstew** [comment]
I've incorporated those bits and pieces. Also the test failures should be fixed now, so we get a green light on travis.

File: verilator-config.cmake.in [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+  PROPERTY VERILATOR
+  BRIEF_DOCS "Verilator"
+  FULL_DOCS "Verilator"
+)
+
+define_property(TARGET
+  PROPERTY VERILATOR_PREFIXES
+  BRIEF_DOCS "List of Verilated prefixes added to this target"
+  FULL_DOCS "List of Verilated prefixes added to this target"
+)
```
Comment (patstew) [inline verilator-config.cmake.in]: @madebr `VERILATOR` and `VERILATOR_PREFIXES` seem unnecessary, do you need the for something? `VERILATOR_PREFIXES` doesn't seem to be used anywhere, and we could just always do `target_link_libraries(${TARGET} PUBLIC ${VERILATOR_MT_CFLAGS})`

Comment (madebr) [reply verilator-config.cmake.in]: Yup, they are not really needed per se.
They are more of a hint to users that this targets contains VERILATED modules.

File: verilator-config.cmake.in [hunk 2]
```diff
@@ -261,11 +244,13 @@ function(verilate TARGET)
 
   if (${VERILATE_PREFIX}_TRACE_VCD)
     # If any verilate() call specifies TRACE, define VM_TRACE in the final build
+    set_property(TARGET ${TARGET} PROPERTY VERILATOR_TRACE ON)
```
Comment (madebr) [inline verilator-config.cmake.in]: There is no `VERILATOR_TRACE` property defined.

File: verilator-config.cmake.in [hunk 3]
```diff
@@ -261,11 +244,13 @@ function(verilate TARGET)
...
     set_property(TARGET ${TARGET} PROPERTY VERILATOR_TRACE_VCD ON)
   endif()
 
   if (${VERILATE_PREFIX}_TRACE_FST)
     # If any verilate() call specifies TRACE_FST, define VM_TRACE_FST in the final build
+    set_property(TARGET ${TARGET} PROPERTY VERILATOR_TRACE ON)
```
Comment (madebr) [inline verilator-config.cmake.in]: There is no `VERILATOR_TRACE` property defined.

File: verilator-config.cmake.in [hunk 4]
```diff
@@ -181,9 +163,14 @@ function(verilate TARGET)
...
   set(VDIR "${BINARY_DIR}/CMakeFiles/${TARGET_NAME}.dir/${VERILATE_PREFIX}.dir")
+
+  if (VERILATE_DIRECTORY)
+    set(VDIR "${VERILATE_DIRECTORY}")
+  endif()
+
   file(MAKE_DIRECTORY ${VDIR})
 
-  set(VERILATOR_COMMAND "${VERILATOR_BIN}" -Wall -Wno-fatal --compiler ${COMPILER}
+  set(VERILATOR_COMMAND "${VERILATOR_BIN}" --compiler ${COMPILER}
```
Comment (madebr) [inline verilator-config.cmake.in]: I'm not sure. Will this work with spaces in `VERILATE_PREFIX` or `VDIR`?

File: verilator-config.cmake.in (hunk 3 above)
Comment (patstew) [reply verilator-config.cmake.in]: Properties don't need to be defined, it's just for documentation. And we don't actually want users to edit those properties. I'll add it anyway for completeness.

File: verilator-config.cmake.in (hunk 4 above)
Comment (patstew) [reply verilator-config.cmake.in]: Spaces in VERILATE_PREFIX will never work, we create variables like `${VERILATE_PREFIX}_THREADS`. I'll check the quoting of VDIR.

Comment (madebr) [reply verilator-config.cmake.in]: About `VERILATE_PREFIX`, you're correct, but when there are no quotes, the error message might be different and confusing (unknown argument).

Comment (patstew) [reply verilator-config.cmake.in]: Actually, this is not a problem. Spaces work as-is for VDIR, just tested. You need to put quotes in the verilate() command for it to be interpreted as a single argument, and those quotes carry through. Prefix also gets passed through 'correctly' at which point verilator generates class names with spaces in...

**Comment by wsnyder** [comment]
Thanks for all this. Code pushed to master towards 4.022.  See bug1363.

**Pull request closed without merging**